src/gd_avif.c

```diff
@@ -0,0 +1,677 @@
+
+	Returns:
+
+		* for <gdImageAvifEx>, <gdImageAvif>, and <gdImageAvifCtx>, nothing.
+		* for <gdImageAvifPtrEx> and <gdImageAvifPtr>, a pointer to the image in memory.
+*/
+
+/*
+	 Function: _gdImageAvifCtx
+
+	 We need this underscored function because gdImageAvifCtx() can't return anything.
+	 And our functions that operate on a memory buffer need to know whether the encoding has succeeded.
+
+	 If we're passed the QUALITY_DEFAULT of -1, set the quantizer params to QUANTIZER_DEFAULT.
+
+	 This function returns 0 on success, or 1 on failure.
+ */
+static avifBool _gdImageAvifCtx(gdImagePtr im, gdIOCtx *outfile, int quality, int speed)
+{
+	avifResult result;
+	avifRGBImage rgb;
+	avifRWData avifOutput = AVIF_DATA_EMPTY;
+	avifBool failed = AVIF_FALSE;
+	avifBool lossless = quality == 100;
+	avifEncoder *encoder = NULL;
+
+	uint32_t val;
+	uint8_t *p;
+	uint8_t a;
+	int x, y;
+
+	if (im == NULL)
+		return 1;
+
+	if (!gdImageTrueColor(im)) {
+		gd_error("avif doesn't support palette images");
+		return 1;
+	}
+
+	if (!gdImageSX(im) || !gdImageSY(im)) {
+		gd_error("image dimensions must not be zero");
+		return 1;
+	}
+
+	if (overflow2(gdImageSX(im), gdImageSY(im))) {
+		gd_error("image dimensions are too large");
+		return 1;
+	}
+
+	if (speed != AVIF_SPEED_DEFAULT)
+		speed = CLAMP(speed, AVIF_SPEED_SLOWEST, AVIF_SPEED_FASTEST);
+
+	avifPixelFormat subsampling = quality >= HIGH_QUALITY_SUBSAMPLING_THRESHOLD ?
+                                CHROMA_SUBAMPLING_HIGH_QUALITY : CHROMA_SUBSAMPLING_DEFAULT;
+
+	// Create the AVIF image.
+	// Set the ICC to sRGB, as that's what gd supports right now.
+	// Note that MATRIX_COEFFICIENTS_IDENTITY enables lossless conversion from RGB to YUV.
+
+	avifImage *avifIm = avifImageCreate(gdImageSX(im), gdImageSY(im), 8, subsampling);
+
+	avifIm->colorPrimaries = AVIF_COLOR_PRIMARIES_BT709;
+	avifIm->transferCharacteristics = AVIF_TRANSFER_CHARACTERISTICS_SRGB;
+	avifIm->matrixCoefficients = lossless ? AVIF_MATRIX_COEFFICIENTS_IDENTITY : AVIF_MATRIX_COEFFICIENTS_BT709;
+
+	avifRGBImageSetDefaults(&rgb, avifIm);
+	avifRGBImageAllocatePixels(&rgb); // this allocates memory, and sets rgb.rowBytes and rgb.pixels.
+
+	// Parse RGB data from the GD image, and copy it into the AVIF RGB image.
+	// Convert 7-bit GD alpha channel values to 8-bit AVIF values.
+
+	p = rgb.pixels;
+	for (y = 0; y < rgb.height; y++) {
+		for (x = 0; x < rgb.width; x++) {
+			val = im->tpixels[y][x];
+
+			*(p++) = gdTrueColorGetRed(val);
+			*(p++) = gdTrueColorGetGreen(val);
+			*(p++) = gdTrueColorGetBlue(val);
+			*(p++) = alpha7BitTo8Bit(gdTrueColorGetAlpha(val));
+		}
+	}
+
+	// Convert the RGB image to YUV.
+
+	result = avifImageRGBToYUV(avifIm, &rgb);
+	failed = isAvifError(result, "Could not convert image to YUV");
+	if (failed)
+		goto cleanup;
+
+	// Encode the image in AVIF format.
+
+	encoder = avifEncoderCreate();
+	int quantizerQuality = quality == QUALITY_DEFAULT ?
+												 QUANTIZER_DEFAULT : quality2Quantizer(quality);
```

Comment:
bad indentation

Reply:
is this better? I just indented the second line by a single tab, instead of trying to line up the second line horizontally with the word "quality".

Reply:
And I just made the same change a few lines earlier:

```c
	avifPixelFormat subsampling = quality >= HIGH_QUALITY_SUBSAMPLING_THRESHOLD ?
		CHROMA_SUBAMPLING_HIGH_QUALITY : CHROMA_SUBSAMPLING_DEFAULT;
```